### PR TITLE
[v2] Migration guide updates

### PR DESCRIPTION
**Pull request by m-allanson**
Build on @ryanditjia's previous work, adding some more sections to the migration guide.

I've left several TODO notes in there, and there's a note in #3986 about expanding the layouts section for site with many pages.

I suspect this will need a few rounds of revisions, but this at least covers every item mentioned in [the breaking changes doc](https://github.com/gatsbyjs/gatsby/blob/v2/Breaking%20Changes.md).

Deploy preview: https://deploy-preview-5661--gatsbyjs.netlify.com/docs/migrating-from-v1-to-v2/

Refs: #3986

### CODE REVIEW

File: docs/docs/migrating-from-v1-to-v2.md [hunk 1]
```diff
@@ -180,17 +168,80 @@ export default props => (
...
+`
+```
+
+Further examples can be found in the [Gatsby Image docs](https://github.com/gatsbyjs/gatsby/tree/d0e29272ed7b009dae18d35d41a45e700cdcab0d/packages/gatsby-image).
+
+## Manually specify postcss plugins
+
+Gatsby v2 removed `postcss-cssnext` and `postcss-import` from the default postcss setup.
+
+Use `modifyWebpackConfig` to specify your postcss plugins.
```
Comment (KyleAMathews) [inline docs/docs/migrating-from-v1-to-v2.md]: Link this to the new webpack docs page

File: docs/docs/migrating-from-v1-to-v2.md [hunk 2]
```diff
@@ -175,22 +179,203 @@ export default props => (
...
+
+## Manually specify postcss plugins
+
+Gatsby v2 removed `postcss-cssnext` and `postcss-import` from the default postcss setup.
+
+Use [`onCreateWebpackConfig`](/docs/add-custom-webpack-config) to specify your postcss plugins.
+
+Note: there will be a `postcss` plugin that allows you to configure postcss from a standard postcss config file. [Follow this discussion on issue 3284](https://github.com/gatsbyjs/gatsby/issues/3284).
+
+## Files mixing CommonJS with ES6 modules won't compile
```
Comment (calcsam) [inline docs/docs/migrating-from-v1-to-v2.md]: Can we change this to something declarative, eg -- "Convert to either pure CommonJS or pure ES6"?

File: docs/docs/migrating-from-v1-to-v2.md [hunk 3]
```diff
@@ -175,22 +179,203 @@ export default props => (
...
+
+Mixing `import` and `module.exports` is not ok:
+```js
+import foo from "foo"
+module.exports = foo;
+```
+
+See [Gatsby's babel docs for more details](/docs/babel).
+
+## Don't query nodes by ID
```
Comment (calcsam) [inline docs/docs/migrating-from-v1-to-v2.md]: Perhaps eg "Rewrite node ID queries"?

File: docs/docs/migrating-from-v1-to-v2.md [hunk 4]
```diff
@@ -175,22 +179,203 @@ export default props => (
...
 
 ## Rename `boundActionCreators` to `actions`
+
 `boundActionCreators` is deprecated in v2. You can continue using it, but it’s recommended that you rename it to `actions`.
 
+> TODO: document new actions - see [actions](/docs/actions)
+
 ## Rename `pathContext` to `pageContext`
+
 Similar to `boundActionCreators` above, `pathContext` is deprecated in favor of `pageContext`.
```
Comment (calcsam) [inline docs/docs/migrating-from-v1-to-v2.md]: Isn't this a breaking change? If so we should highlight that -- I don't believe boundActionCreators is.

File: docs/docs/migrating-from-v1-to-v2.md [hunk 5]
```diff
@@ -175,22 +179,203 @@ export default props => (
...
+      })
+  }
+}
+```
+
+Note usage of the new [`setWebpackConfig` action](/docs/actions/#setWebpackConfig).
+
+See [Gatsby's webpack docs for more details](/docs/add-custom-webpack-config) about configuring webpack.
+
+## CSS is inlined automatically
```
Comment (calcsam) [inline docs/docs/migrating-from-v1-to-v2.md]: Perhaps "Remove inlined css in html.js"?

File: docs/docs/migrating-from-v1-to-v2.md [hunk 6]
```diff
@@ -175,22 +179,203 @@ export default props => (
...
+
+See [Gatsby's babel docs for more details](/docs/babel).
+
+## Don't query nodes by ID
+
+Source and transformer plugins now use UUIDs for IDs. If you used glob or regex to query nodes by id then you'll need to query something else.
+
+[See the Pull Request that implemented this change](https://github.com/gatsbyjs/gatsby/pull/3807/files)
+
+> TODO: add example
```
Comment (jlengstorf) [inline docs/docs/migrating-from-v1-to-v2.md]: This is a real example from one of my projects. 😄 

```diff
  query MyImageQuery {
    allImageSharp(filter: {
-     id: {regex: "/default.jpg/"}
+     sizes: {originalName: {regex: "/default.jpg/"}}
    }) {
      edges {
        node {
          id
          sizes(maxWidth: 660) {
            src
          }
        }
      }
    }
  }
```

File: docs/docs/migrating-from-v1-to-v2.md [hunk 7]
```diff
@@ -175,22 +179,203 @@ export default props => (
...
+
+## Remove explicit polyfills
+
+If your Gatsby v1 site included any polyfills, you can remove them. Gatsby v2 ships with babel 7 and is configured to automatically include polyfills for your code. See [Gatsby's babel docs for more details](/docs/babel).
```
Comment (jlengstorf) [inline docs/docs/migrating-from-v1-to-v2.md]: This doesn't appear to be actually true, or I'm misunderstanding. I had to ship an IntersectionObserver polyfill for marisamorby.com:

https://github.com/jlengstorf/marisamorby.com/pull/8/files#diff-3863fe9c963d5c58d29a7f5af77e6480

File: docs/docs/migrating-from-v1-to-v2.md [hunk 8]
```diff
@@ -175,22 +179,203 @@ export default props => (
...
+
+## Change `modifyWebpackConfig` to `onCreateWebpackConfig`
+
+`modifyWebpackConfig` was renamed to [`onCreateWebpackConfig`](/docs/node-apis/#onCreateWebpackConfig) to bring it in line with the rest of Gatsby's API names.
+
+Before:
+
+```js
+exports.modifyWebpackConfig = ({ config, stage }) => {
+  switch (stage) {
```
Comment (jlengstorf) [inline docs/docs/migrating-from-v1-to-v2.md]: 😱 a _switch statement?!_ 🤣

File: docs/docs/migrating-from-v1-to-v2.md (hunk 4 above)
Comment (KyleAMathews) [reply docs/docs/migrating-from-v1-to-v2.md]: It's not — we pass both `pathContext` and `pageContext` as props

File: docs/docs/migrating-from-v1-to-v2.md (hunk 3 above)
Comment (KyleAMathews) [reply docs/docs/migrating-from-v1-to-v2.md]: Hmmm I think @m-allanson original wording is correct. There aren't "node ID queries" just queries that might try to query nodes by ID.

File: docs/docs/migrating-from-v1-to-v2.md (hunk 5 above)
Comment (KyleAMathews) [reply docs/docs/migrating-from-v1-to-v2.md]: 👍

File: docs/docs/migrating-from-v1-to-v2.md [hunk 9]
```diff
@@ -51,7 +50,7 @@ export default ({ children }) => (
 )
 ```
 
-`layout in v2`
+After:
 
 ```jsx{5}
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply docs/docs/migrating-from-v1-to-v2.md]: 👍

File: docs/docs/migrating-from-v1-to-v2.md [hunk 10]
```diff
@@ -175,22 +179,203 @@ export default props => (
...
+CommonJS is ok:
+
+```js
+const foo = require('foo');
+module.exports = foo;
+```
+
+Mixing `requires` and `export` is not ok:
+```js
+const foo = require('foo');
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply docs/docs/migrating-from-v1-to-v2.md]: 👍 

The Eslint docs do this too and it's really nice

File: docs/docs/migrating-from-v1-to-v2.md (hunk 7 above)
Comment (KyleAMathews) [reply docs/docs/migrating-from-v1-to-v2.md]: Gatsby isn't yet compiling code in `node_modules` so code there won't be polyfilled yet. See bullet number 5 https://github.com/gatsbyjs/gatsby/issues/3870

File: docs/docs/migrating-from-v1-to-v2.md [hunk 11]
```diff
@@ -87,10 +88,12 @@ export default () => (
...
+
 Since layout is no longer special, you now need to make use of v2’s StaticQuery feature.
 
-`Query with layout in v1`
+> TODO: document StaticQuery and link to it from here
+
+Before:
 
 ```jsx
 import React, { Fragment } from "react"
```
[the remark this answers is not in the document]
Comment (m-allanson) [reply docs/docs/migrating-from-v1-to-v2.md]: Nitpicky is good :)

File: docs/docs/migrating-from-v1-to-v2.md (hunk 6 above)
Comment (m-allanson) [reply docs/docs/migrating-from-v1-to-v2.md]: Thanks :)

File: docs/docs/migrating-from-v1-to-v2.md (hunk 8 above)
Comment (m-allanson) [reply docs/docs/migrating-from-v1-to-v2.md]: hah! I'll have you know this was copy and pasted from an official Gatsby plugin 🤣

**Comment by m-allanson** [comment]
Thanks everyone for the top notch feedback, this is ready for another look.

Everything should be addressed with the exception of the [polyfill section](https://github.com/gatsbyjs/gatsby/pull/5661#discussion_r192534484), where I've just added a disclaimer to say that this doesn't yet work for code imported from `node_modules`. I'm slightly wary of promising features that don't yet exist, so could edit that section.

**Comment by m-allanson** [comment]
I added one more section on how to install v2 prereleases (thanks for the idea @pieh)

**Comment by m-allanson** [comment]
I think this is good enough to merge for now, but wonder if it could be divided out into more useful sections. I'm not sure what that would look like - maybe just one section for breaking changes and one for deprecations.

**Comment by m-allanson** [comment]
I'm going to merge this, we can do further changes as additional PRs. Thanks all!

**Pull request merged**